### ISSUE TITLE
OpenSuse Tumbleweed - azure-cli-telemetry doesn't support python311-portalocker in version 3 and above

### ISSUE THREAD

**Issue by martinspudich** [Azure/azure-cli#30458]
### Describe the bug

When I'm trying to update my system via `sudo zypper dup` I'm getting an error with python311-portalocker version issue.

Please update azure-cli-telemetry to work with azure-cli-telemetry version 3 and above.

### Related command

sudo zypper dup
Loading repository data...
Reading installed packages...
Warning: You are about to do a distribution upgrade with all enabled repositories. Make sure these repositories are compatible before you continue. See 'man zypper' for more information about this command.
Computing distribution upgrade...

Problem: 1: the installed azure-cli-telemetry-1.1.0-3.1.noarch requires 'python311-portalocker < 3.0', but this requirement cannot be provided
deleted providers: python311-portalocker-2.10.1-1.2.noarch

 Solution 1: Following actions will be done:
  deinstallation of azure-cli-telemetry-1.1.0-3.1.noarch
  deinstallation of azure-cli-core-2.66.0-1.1.noarch
  deinstallation of azure-cli-2.66.0-1.1.noarch
 Solution 2: keep obsolete python311-portalocker-2.10.1-1.2.noarch
 Solution 3: break azure-cli-telemetry-1.1.0-3.1.noarch by ignoring some of its dependencies

### Errors

Problem: 1: the installed azure-cli-telemetry-1.1.0-3.1.noarch requires 'python311-portalocker < 3.0', but this requirement cannot be provided
deleted providers: python311-portalocker-2.10.1-1.2.noarch

### Issue script & Debug output

sudo zypper dup
Loading repository data...
Reading installed packages...
Warning: You are about to do a distribution upgrade with all enabled repositories. Make sure these repositories are compatible before you continue. See 'man zypper' for more information about this command.
Computing distribution upgrade...

Problem: 1: the installed azure-cli-telemetry-1.1.0-3.1.noarch requires 'python311-portalocker < 3.0', but this requirement cannot be provided
deleted providers: python311-portalocker-2.10.1-1.2.noarch

 Solution 1: Following actions will be done:
  deinstallation of azure-cli-telemetry-1.1.0-3.1.noarch
  deinstallation of azure-cli-core-2.66.0-1.1.noarch
  deinstallation of azure-cli-2.66.0-1.1.noarch
 Solution 2: keep obsolete python311-portalocker-2.10.1-1.2.noarch
 Solution 3: break azure-cli-telemetry-1.1.0-3.1.noarch by ignoring some of its dependencies

### Expected behavior

Resolving the package version conflict.

### Environment Summary

 az --version
azure-cli                         2.66.0 *

core                              2.66.0 *
telemetry                          1.1.0

Dependencies:
msal                              1.31.0
azure-mgmt-resource               23.2.0

Python location '/usr/bin/python3.11'
Extensions directory '/home/bliker/.azure/cliextensions'

Python (Linux) 3.11.10 (main, Sep 09 2024, 17:03:08) [GCC]

Legal docs and information: aka.ms/AzureCliLegal

**Comment by yonzhan** [comment]
Thank you for opening this issue, we will look into it.

**Comment by evelyn-ys** [comment]
cc @jiasl@rayluo for awareness. MSAL is also restricting `portalocker<3`

**Comment by bebound** [comment]
I've used https://github.com/AzureAD/microsoft-authentication-extensions-for-python/pull/117 in CLI on SAW 2.55.0 and it works.

**Comment by rayluo** [comment]
> I've used [AzureAD/microsoft-authentication-extensions-for-python#117](https://github.com/AzureAD/microsoft-authentication-extensions-for-python/pull/117) in CLI on SAW 2.55.0 and it works.

@bebound Thanks for the promising feedback! You also brought up a good point on SAW. I am not even sure whether the portalocker was allowed on SAW, and if not, how the previous versions of AzureCLI + MsalExtensions + Portalocker combination worked (or, did they?). Regardless, this becomes yet another good reason to trim down that dependency.

@jiasli , do we want this change to be in one of [upcoming Azure CLI releases](https://github.com/Azure/azure-cli/milestones)?

**Comment by rayluo** [comment]
> I've used this PR's msal-extension without portalocker since the initial release of SAW version 2.40.0.

@bebound, when was that [Azure CLI 2.40.0 released? Was it in September 2022?](https://github.com/Azure/azure-cli/releases/tag/azure-cli-2.40.0)

Sorry I don't quite remember our meeting. Did you mean you used/tested that portalocker-less msal-extension for 2+ years already?

**Comment by bebound** [comment]
[CLI on SAW](https://ppe.eng.ms/docs/cloud-ai-platform/azure-core/azure-management-and-platforms/control-plane-bburns/azure-cli-tools-azure-cli-powershell-and-terraform/azure-cli-tools/teams_docs/azcli_docs/azure_cli_on_saw) is an internal version. It's been released for two years, and we haven't received any bug report related to msal. At least it works for the SAW version.

I assume dropping portalocker causes issues in high concurrency environments?

**Comment by rayluo** [comment]
> I assume dropping portalocker causes issues in high concurrency environments?

No. Because that PR does not simply drop portalocker. It also uses a replacement locking mechanism without external dependencies.

> [CLI on SAW](https://ppe.eng.ms/docs/cloud-ai-platform/azure-core/azure-management-and-platforms/control-plane-bburns/azure-cli-tools-azure-cli-powershell-and-terraform/azure-cli-tools/teams_docs/azcli_docs/azure_cli_on_saw) is an internal version. It's been released for two years, and we haven't received any bug report related to msal. At least it works for the SAW version.

Thanks for the much-awaited confirmation! I'll merge that PR in, then.
